src/AWS/Lambda/S3Event.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+module AWS.Lambda.S3Event
+  (
+    S3Event(..),
+    S3(..),
+    S3Object(..),
+    Bucket(..),
+    Record(..)
+  ) where
+
+import Data.Aeson
+import Data.HashMap.Strict
+import Data.Text (Text)
+import GHC.Generics
+
+data S3 a =
+  S3 {
+    s3SchemaVersion :: Text,
+    configurationId :: Text,
+    bucket          :: Bucket,
+    s3Object        :: S3Object
+  } deriving (Show)
+
+instance FromJSON a => FromJSON (S3 a) where
+  parseJSON (Object v) = S3
+    <$> v .: "s3SchemaVersion"
+    <*> v .: "configurationId"
+    <*> v .: "bucket"
+    <*> ((v .: "object") >>= parseJSON)
+
+instance ToJSON a => ToJSON (S3 a) where
+  toJSON S3{..} =
+    object [
+        "s3SchemaVersion" .= s3SchemaVersion
+      , "configurationId" .= configurationId
+      , "bucket" .= bucket
+      , "object" .= s3Object
+    ]
+
+data Bucket =
+  Bucket {
+    name          :: Text,
+    arn           :: Text,
+    ownerIdentity :: Maybe (HashMap Text Text)
+  } deriving (Show, Generic)
+
+instance FromJSON Bucket
+instance ToJSON Bucket
+
+data S3Object =
+  S3Object {
+    key       :: Text,
+    size      :: Maybe (Integer),
+    eTag      :: Maybe (Text),
+    versionId :: Text,
```

Comment:
```suggestion
    versionId :: Maybe Text,
```
I tested this with sam generated events (`sam local generate-event s3 put`) and looks like versionId is optional